planner/core/rule_partition_processor.go

```diff
@@ -224,73 +224,251 @@ func (s *partitionProcessor) processHashPartition(ds *DataSource, pi *model.Part
+}
+
+func (l *listPartitionPruner) locatePartitionByCNFCondition(conds []expression.Expression) (tables.ListPartitionLocation, bool, error) {
+	if len(conds) == 0 {
+		return nil, true, nil
+	}
+	countFull := 0
+	intersect := tables.NewListPartitionLocationIntersect()
+	for _, cond := range conds {
+		cnfLoc, isFull, err := l.locatePartition(cond)
+		if err != nil {
+			return nil, false, err
+		}
+		if isFull {
+			countFull++
+			continue
+		}
+		if cnfLoc.IsEmpty() {
+			// No partition for intersection, just return 0 partition.
+			return nil, false, nil
+		}
+		if !intersect.Intersect(cnfLoc) {
+			return nil, false, nil
+		}
+	}
+	if countFull == len(conds) {
+		return nil, true, nil
+	}
+	return intersect.GetLocation(), false, nil
+}
+
+func (l *listPartitionPruner) locatePartitionByDNFCondition(conds []expression.Expression) (tables.ListPartitionLocation, bool, error) {
+	if len(conds) == 0 {
+		return nil, true, nil
+	}
+	merger := tables.NewListPartitionLocationMerger()
+	for _, cond := range conds {
+		dnfLoc, isFull, err := l.locatePartition(cond)
+		if err != nil || isFull {
+			return nil, isFull, err
+		}
+		merger.Merge(dnfLoc)
+	}
+	return merger.GetLocation(), false, nil
+}
+
+// locatePartitionByColumn uses to locate partition by the one of the list columns value.
+// Such as: partition by list columns(a,b) (partition p0 values in ((1,1),(2,2)), partition p1 values in ((6,6),(7,7)));
+// and if the condition is `a=1`, then we can use `a=1` and the expression `(a in (1,2))` to locate partition `p0`.
+func (l *listPartitionPruner) locatePartitionByColumn(cond *expression.ScalarFunction) (tables.ListPartitionLocation, bool, error) {
+	condCols := expression.ExtractColumns(cond)
+	if len(condCols) != 1 {
+		return nil, true, nil
+	}
+	var colPrune *tables.ForListColumnPruning
+	for _, cp := range l.listPrune.ColPrunes {
+		if cp.ExprCol.ID == condCols[0].ID {
+			colPrune = cp
+		}
+	}
+	if colPrune == nil {
+		return nil, true, nil
+	}
+	return l.locateColumnPartitionsByCondition(cond, colPrune)
+}
+
+func (l *listPartitionPruner) locateColumnPartitionsByCondition(cond expression.Expression, colPrune *tables.ForListColumnPruning) (tables.ListPartitionLocation, bool, error) {
+	ranges, err := l.detachCondAndBuildRange([]expression.Expression{cond}, colPrune.ExprCol)
+	if err != nil {
+		return nil, false, err
+	}
+
+	sc := l.ctx.GetSessionVars().StmtCtx
+	merger := tables.NewListPartitionLocationMerger()
+	for _, r := range ranges {
+		if r.IsPointNullable(sc) {
+			if len(r.HighVal) != 1 {
+				return nil, true, nil
+			}
+			location, err := colPrune.LocatePartition(sc, r.HighVal[0])
+			if err != nil {
+				return nil, false, err
+			}
+			if len(l.partitionNames) > 0 {
+				for _, pg := range location {
+					if l.findByName(l.partitionNames, l.pi.Definitions[pg.PartIdx].Name.L) {
+						merger.MergePartitionGroup(pg)
+					}
+				}
+			} else {
+				merger.Merge(location)
+			}
+		} else {
+			return nil, true, nil
```

Comment:
maybe query condition like  `a > 1 and a < 3` is common in list partition~?

it seems it's hard to use map structure to support range in future..

> ps: support `b = ?` in  `partition by (a, b)` prune by to level map is cool, but I'm not sure whether it's more useful than range~?

Reply:
Maybe we can also build a range pruner for list partition to handle those range queries?